### ISSUE TITLE
cannot parse reply from rspamd

### ISSUE THREAD

**Issue by kristous** [vstakhov/rspamd#932]
Starting with rspamd_1.3.5-1 on debian jessie I get the following errors on some messages.

A downgrade to rspamd_1.3.4-1 makes the error go away.
That's why I suspect it a rspamd bug and not a rmilter bug.

I don't get the error on ubuntu trusty.

Sep  7 13:02:57 mail1 rmilter[5541]: <600cd128bb>; cannot parse reply from rspamd: no default metric result
Sep  7 13:02:57 mail1 rmilter[5541]: <600cd128bb>; rspamd; HTTP parser error: the on_body callback failed when rspamd reply
Sep  7 13:02:57 mail1 rmilter[5541]: <600cd128bb>; spamdscan: finish scanning message on localhost
Sep  7 13:02:57 mail1 rmilter[5541]: <600cd128bb>; rspamdscan: retry limit exceeded, localhost, /var/spool/rmilter/msg.XXpIW961
Sep  7 13:02:57 mail1 rmilter[5541]: <600cd128bb>; mlfi_eom: spamdscan() failed

**Comment by fatalbanana** [comment]
Message returned by rspamd in such cases may be of some interest (You could collect it with `tcpflow` for example).

**Comment by fatalbanana** [comment]
That is a (expected) message from the _controller_. For scanning (ie. interaction with rmilter) you are interested in normal worker port: 11333. I'm assuming that you did not in fact get this message in response to a scan? (tcpflow records the other half of the conversation in a different file)

**Comment by kristous** [comment]
This was port 11333 and it was the response to scan which caused the error in the logs.

The log of rspamd of this scan:

```
Sep  7 14:53:43 mail1 rspamd[9054]: (normal) <2c21b2>; task; accept_socket: accepted connection from ::1 port 57381, task ptr: 00007F2BE18D05A0
Sep  7 14:53:43 mail1 rspamd[9054]: (normal) <2c21b2>; task; rspamd_message_parse: loaded message; id: <a40ac3c5488b58051415d6d059ea5d4f@cburial.com>; queue-id: <E2E6CC2>; size: 3941; checksum: <25411602f29b7cf9d38e2d1d9fcd9a19>
Sep  7 14:53:43 mail1 rspamd[9054]: (normal) <2c21b2>; task; process_dns_results: <a40ac3c5488b58051415d6d059ea5d4f@cburial.com> domain [cburial.com.dbl.spamhaus.org] is in surbl dbl.spamhaus.org(201007f)
Sep  7 14:53:43 mail1 rspamd[9054]: (normal) <2c21b2>; task; process_dns_results: <a40ac3c5488b58051415d6d059ea5d4f@cburial.com> domain [cburial.com] is in surbl sbl.spamhaus.org(200007f)
Sep  7 14:53:44 mail1 rspamd[9054]: (normal) <2c21b2>; task; bayes_classify: skip classification as ham class has not enough learns: 0, 200 required
Sep  7 14:53:44 mail1 rspamd[9054]: (normal) <2c21b2>; task; rspamd_stat_check_autolearn: <a40ac3c5488b58051415d6d059ea5d4f@cburial.com>: autolearn spam for classifier 'bayes' as message's action is reject, score: 35.10
```

"ham class has not enough learns" sound strange to me because in redis there are  394451

**Comment by kristous** [comment]
I am just checking if statistic.conf syntax has been changed and my override file is not valid any more.
Hold on.

**Comment by vstakhov** [comment]
There were no changes in statistics since 1.3.4.

**Comment by kristous** [comment]
My changes to the statistics override file did not make any changes.
The error happened 5 times out of 10000 spamdscans in the last 12 hours so maybe it is just some kind of overload issue?
The strange thing is, a downgrade to 1.3.4 makes it go away.

**Comment by fatalbanana** [comment]
So I suppose issue relates to minimum learn count being fixed (and autolearning?). I was unable to replicate it however.

**Comment by kristous** [comment]
I think I fixed the "cannot parse reply" errors.
I moved my statistic.conf from the local.d to the override.d and changed the content to what is in the documentation rather than copying /etc/rspamd/statistic.conf and modifying it.

**Comment by qk4l** [comment]
I got the same problem. Rspmac returns 504 error and MTA can not process it.
Setting autolearn to False fix a problem but there is something wrong here.
**rspamc stat**
```
Messages scanned: 71458
Messages with action reject: 348, 0.48%
Messages with action soft reject: 32115, 44.94%
Messages with action rewrite subject: 0, 0.00%
Messages with action add header: 510, 0.71%
Messages with action greylist: 9074, 12.69%
Messages with action no action: 29411, 41.15%
Messages treated as spam: 32973, 46.14%
Messages treated as ham: 38485, 53.85%
Messages learned: 528
Connections count: 26
Control connections count: 1
Pools allocated: 97
Pools freed: 58
Bytes allocated: 670k
Memory chunks allocated: 65
Shared chunks allocated: 11
Chunks freed: 0
Oversized chunks: 1
Fuzzy hashes in storage "example.com": 2586
Fuzzy hashes in storage "rspamd.com": 768107
Fuzzy hashes stored: 770693
Statfile: BAYES_HAM type: redis; length: 31.92M; free blocks: 0; total blocks: 840k; free: 0.00%; learned: 1642; users: 1; languages: 0
Statfile: BAYES_SPAM type: redis; length: 29.59M; free blocks: 0; total blocks: 778k; free: 0.00%; learned: 4534; users: 1; languages: 0
Total learns: 0
```
classifirer block from `configdump`
```
classifier {
    bayes {
        tokenizer {
            name = "osb";
        }
        name = "bayes";
        min_tokens = 11;
        min_learns = 200;
        backend = "redis";
        database = 2;
        servers = "master-slave: ro0.exampe.com, ro1.example.com";
        write_servers = "rw.example.com";
        password = "REDIS PASSWORD";
        statfile {
            symbol = "BAYES_SPAM";
        }
        statfile {
            symbol = "BAYES_HAM";
        }
        autolearn = True;
    }
}
```

**Comment by qk4l** [comment]
@fatalbanana may be the problem that rspamd incorrect count learns from Redis database?

> skip classification as ham class has not enough learns: 0, 200 required

`rspamc stat` also shows `Total learns: 0` while I have more than 1000.

**Comment by vstakhov** [comment]
This issue has been recently fixed in the master. Nevertheless, learn errors should not be propagated to the scanner when it is learned implicitly (using autolearn).

**Comment by qk4l** [comment]
@vstakhov Thank you for information. I'm waiting next release and temporary disable autolearn process

**Comment by vstakhov** [comment]
I don't think I'll backport these changes to 1.3 branch. I'm looking forward towards 1.4 release and have no time to maintain other branches since changes are too huge. On the other hand, 1.4 is pretty stable now.

**Comment by vstakhov** [comment]
Ok, it seems I know the reason. Redis cache tries to set error for a task despite of the fact that it is autolearn. It is fixed in 203de2b.
